File: app/src/main/java/ir/logcat/canvastext/view/AutoFitTextView.java
package ir.logcat.canvastext.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.support.v4.content.ContextCompat;
import android.text.TextPaint;
import android.view.View;

import ir.logcat.canvastext.R;
import ir.logcat.canvastext.utils.Utils;

public class AutoFitTextView extends View {

    private TextPaint textPaint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
    private String text = "لورم ایپسوم متن ساختگی";
    private int leftMargin, topMargin, width, height, padding;
    private Rect backgroundRect = new Rect();
    private Paint backgroundPaint = new Paint();

    public AutoFitTextView(Context context) {
        super(context);
        init();
    }

    private void init() {

        textPaint.setColor(ContextCompat.getColor(getContext(), R.color.white));
        textPaint.setTextAlign(Paint.Align.CENTER);
        textPaint.setTypeface(Typeface.createFromAsset(getContext().getAssets(), "fonts/iran_sans_light.ttf"));

        leftMargin = topMargin = Utils.dpToPx(getContext(), 32);
        width = Utils.dpToPx(getContext(), 150);
        height = Utils.dpToPx(getContext(), 50);
        padding = Utils.dpToPx(getContext(), 16);


        backgroundRect.set(leftMargin, topMargin, leftMargin + width, topMargin + height);

        backgroundPaint.setStyle(Paint.Style.FILL_AND_STROKE);
        backgroundPaint.setColor(ContextCompat.getColor(getContext(), R.color.colorAccent));

        int defaultTextSize = Utils.spToPx(getContext(), 20);
        textPaint.setTextSize(defaultTextSize);
        float desiredTextSize = defaultTextSize * backgroundRect.width() / (textPaint.measureText(text) + padding * 2);
        textPaint.setTextSize(desiredTextSize);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.drawRect(backgroundRect, backgroundPaint);
        canvas.drawText(text, backgroundRect.centerX(), backgroundRect.centerY() - (textPaint.ascent() + textPaint.descent()) / 2, textPaint);
    }
}

File: app/src/main/java/ir/logcat/canvastext/view/kotlin/AutoFitTextView.kt
package ir.logcat.canvastext.view.kotlin

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.Typeface
import android.support.v4.content.ContextCompat
import android.text.TextPaint
import android.view.View
import ir.logcat.canvastext.R
import ir.logcat.canvastext.utils.Utils

class AutoFitTextView(context: Context) : View(context) {

    private val textPaint = TextPaint(Paint.ANTI_ALIAS_FLAG)
    private val text = "لورم ایپسوم متن ساختگی"
    private var leftMargin = 0
    private var topMargin = 0
    private var w = 0
    private var h = 0
    private var padding = 0
    private val backgroundRect = Rect()
    private val backgroundPaint = Paint()

    init {

        textPaint.color = ContextCompat.getColor(context, R.color.white)
        textPaint.textAlign = Paint.Align.CENTER
        textPaint.typeface = Typeface.createFromAsset(context.assets, "fonts/iran_sans_light.ttf")

        topMargin = Utils.dpToPx(context, 32)
        leftMargin = topMargin
        w = Utils.dpToPx(context, 150)
        h = Utils.dpToPx(context, 50)
        padding = Utils.dpToPx(context, 16)

        backgroundRect.set(leftMargin, topMargin, leftMargin + w, topMargin + h)

        backgroundPaint.style = Paint.Style.FILL_AND_STROKE
        backgroundPaint.color = ContextCompat.getColor(context, R.color.colorAccent)

        val defaultTextSize = Utils.spToPx(context, 20)
        textPaint.textSize = defaultTextSize.toFloat()
        val desiredTextSize = defaultTextSize * backgroundRect.width() / (textPaint.measureText(text) + padding * 2)
        textPaint.textSize = desiredTextSize
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.drawRect(backgroundRect, backgroundPaint)
        canvas.drawText(text, backgroundRect.centerX().toFloat(), backgroundRect.centerY() - (textPaint.ascent() + textPaint.descent()) / 2, textPaint)
    }
}
